### PR TITLE
refactor: remove $_SESSION from methods and functions

### PR DESCRIPTION
**Pull request by pjsde**
Functions and methods should not directly call the `$_SESSION` array but use the `session()` helper to obtain their contents from the `Services::Session` library.

This avoids the need to override code if someone needs to organize the `$_SESSION` array in another way.

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/CodeIgniter.php [hunk 1]
```diff
@@ -1063,14 +1063,18 @@ public function storePreviousURL($uri)
...
-            );
+        $previousUrl = URI::createURIString(
+            $uri->getScheme(),
+            $uri->getAuthority(),
+            $uri->getPath(),
+            $uri->getQuery(),
+            $uri->getFragment()
+        );
+
+        if (ENVIRONMENT === 'testing') {
```
Comment (kenjis) [inline system/CodeIgniter.php:1074]: Eventually this condition (`ENVIRONMENT === 'testing'`) should be removed in all the code base.
And the Session class should work in testing.
Is there something wrong?

Comment (pjsde) [reply system/CodeIgniter.php:1074]: If you don't have this validation, there are a series of tests that fail loke this:

```
ErrorException : ini_set(): Headers already sent. You cannot change the session module's ini settings at this time
 D:\projects\CodeIgniter4\system\Session\Handlers\FileHandler.php:72
 D:\projects\CodeIgniter4\system\Config\Services.php:686
 D:\projects\CodeIgniter4\system\Config\BaseService.php:258
 D:\projects\CodeIgniter4\system\Config\BaseService.php:199
 D:\projects\CodeIgniter4\system\Config\Services.php:664
 D:\projects\CodeIgniter4\system\Config\BaseService.php:258
 D:\projects\CodeIgniter4\system\Common.php:996
 D:\projects\CodeIgniter4\system\CodeIgniter.php:1075
 D:\projects\CodeIgniter4\system\CodeIgniter.php:549
 D:\projects\CodeIgniter4\system\CodeIgniter.php:361
 D:\projects\CodeIgniter4\tests\system\CodeIgniterTest.php:617
```

I don't know how to use `Session` in tests, but if you could help me I would appreciate it.

Comment (pjsde) [reply system/CodeIgniter.php]: @kenjis i remove the validations you mentioned and needed to update tests to work. i add the clause `@runTestsInSeparateProcesses` to the tests and they run in my machibe whitout errors (windows with apache php 7.4) but here they throw errors. Can you help me to see if how can i fix it please?

Comment (pjsde) [reply system/CodeIgniter.php]: fixed errors

**Review by kenjis — APPROVED** [review]
This is an improvement.
Thank you.

File: system/Helpers/url_helper.php [hunk 2]
```diff
@@ -88,8 +88,12 @@ function previous_url(bool $returnObject = false)
...
         // since it's more reliable and safer.
+        if (isset($_SESSION)) {
+            $referer = session('_ci_previous_url');
+        }
+
         // Otherwise, grab a sanitized version from $_SERVER.
-        $referer = $_SESSION['_ci_previous_url'] ?? Services::request()->getServer('HTTP_REFERER', FILTER_SANITIZE_URL);
+        $referer ??= Services::request()->getServer('HTTP_REFERER', FILTER_SANITIZE_URL);
 
         $referer ??= site_url('/');
```
Comment (mostafakhudair) [inline system/Helpers/url_helper.php:98]:
```suggestion
        $referer ??= request()->getServer('HTTP_REFERER', FILTER_SANITIZE_URL) ?? site_url('/');
```

**Comment by kenjis** [comment]
@pjsde We do not include merge commits in PR branches.
Can you do `git rebase` to remove the merge commit?
See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#updating-your-branch

**Comment by pjsde** [comment]
> @pjsde We do not include merge commits in PR branches. Can you do `git rebase` to remove the merge commit

@kenjis Sorry, I hadn't noticed

**Review by kenjis — APPROVED** [review]

**Comment by kenjis** [comment]
@pjsde Thank you! Glad you are able to use `git rebase`.

**Pull request merged**